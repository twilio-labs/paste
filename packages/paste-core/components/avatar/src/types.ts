import type {IconSize} from '@twilio-paste/style-props';
import type {GenericIconProps} from '@twilio-paste/icons/esm/types';

export interface AvatarProps extends React.HTMLAttributes<'div'> {
  name: string;
  size: IconSize;
  src?: string;
  icon?: React.FC<GenericIconProps>;
}
