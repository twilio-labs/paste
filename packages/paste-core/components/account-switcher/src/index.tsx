export * from "./AccountSwitcher";
export * from "./AccountSwitcherGroup";
export * from "./AccountSwitcherSeparator";
export * from "./AccountSwitcherItem";
export * from "./AccountSwitcherItemRadio";
export * from "./AccountSwitcherBadge";
export * from "./useAccountSwitcher";
