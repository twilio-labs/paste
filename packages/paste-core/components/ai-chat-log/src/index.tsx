export { AIChatMessage } from "./AIChatMessage";
export type { AIChatMessageProps } from "./AIChatMessage";
export { AIChatMessageActionGroup } from "./AIChatMessageActionGroup";
export type { AIChatMessageActionGroupProps } from "./AIChatMessageActionGroup";
export { AIChatMessageBody } from "./AIChatMessageBody";
export type { AIChatMessageBodyProps } from "./AIChatMessageBody";
export { AIChatMessageAuthor } from "./AIChatMessageAuthor";
export type { AIChatMessageAuthorProps } from "./AIChatMessageAuthor";
export { AIChatMessageActionCard } from "./AIChatMessageActionCard";
export type { AIChatMessageActionCardProps } from "./AIChatMessageActionCard";
export { AIChatMessageLoading } from "./AIChatMessageLoading";
export type { AIChatMessageLoadingProps } from "./AIChatMessageLoading";

export { AIChatLog } from "./AIChatLog";
export type { AIChatLogProps } from "./AIChatLog";
export { useAIChatLogger } from "./useAIChatLogger";
export type { UseAIChatLogger, AIPartialIDChat, AIChat } from "./useAIChatLogger";
export { AIChatLogger } from "./AIChatLogger";
export type { AIChatLoggerProps } from "./AIChatLogger";

export type { AIMessageVariants } from "./AIMessageContext";
