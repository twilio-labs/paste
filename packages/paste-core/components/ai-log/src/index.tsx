export { AIChatMessage } from "./AIChatMessage";
export type { AIChatMessageProps } from "./AIChatMessage";
export { AIChatMessageMeta } from "./AIChatMessageMeta";
export type { AIChatMessageMetaProps } from "./AIChatMessageMeta";
export { AIChatMessageMetaItem } from "./AIChatMessageMetaItem";
export type { AIChatMessageMetaItemProps } from "./AIChatMessageMetaItem";
export { AIChatMessageContent } from "./AIChatMessageContent";
export type { AIChatMessageContentProps } from "./AIChatMessageContent";

export { AIChatLog } from "./AIChatLog";
export type { AIChatLogProps } from "./AIChatLog";
export { useAIChatLogger } from "./useAIChatLogger";
export type { UseAIChatLogger } from "./useAIChatLogger";
export { AIChatLogger } from "./AIChatLogger";
export type { AIChatLoggerProps } from "./AIChatLogger";
