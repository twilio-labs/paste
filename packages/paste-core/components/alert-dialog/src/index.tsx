import * as React from 'react';
import * as PropTypes from 'prop-types';
import {useUID} from '@twilio-paste/uid-library';
import {useTransition} from '@twilio-paste/animation-library';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import type {BoxProps} from '@twilio-paste/box';
import {ModalDialogOverlay} from '@twilio-paste/modal';
import {AlertDialogHeader} from './AlertDialogHeader';
import {AlertDialogBody} from './AlertDialogBody';
import {AlertDialogContent} from './AlertDialogContent';
import {AlertDialogFooter} from './AlertDialogFooter';

// eslint-disable-next-line @typescript-eslint/no-explicit-any
const getAnimationStates = (): any => ({
  from: {opacity: 0, transform: `scale(0.675)`},
  enter: {opacity: 1, transform: `scale(1)`},
  leave: {opacity: 0, transform: `scale(0.675)`},
  // https://www.react-spring.io/docs/hooks/api
  config: {
    mass: 0.5,
    tension: 370,
    friction: 26,
  },
});

export interface AlertDialogProps extends React.HTMLAttributes<HTMLElement>, Pick<BoxProps, 'element'> {
  children: NonNullable<React.ReactNode>;
  destructive?: boolean;
  heading: string;
  isOpen: boolean;
  onConfirm: () => void;
  onConfirmLabel: string;
  onDismiss: () => void;
  onDismissLabel: string;
}

export const AlertDialog = React.forwardRef<HTMLDivElement, AlertDialogProps>(
  (
    {
      children,
      destructive,
      element = 'ALERT_DIALOG',
      heading,
      isOpen,
      onConfirm,
      onConfirmLabel,
      onDismiss,
      onDismissLabel,
      ...props
    },
    ref
  ) => {
    const transitions = useTransition(isOpen, getAnimationStates());
    const headingID = useUID();
    const bodyID = useUID();

    return (
      <>
        {transitions(
          (styles, item) =>
            item && (
              <ModalDialogOverlay isOpen={isOpen} style={{opacity: styles.opacity}}>
                <Box
                  // @ts-expect-error Render overlay as box for customization
                  as={AlertDialogContent}
                  {...safelySpreadBoxProps(props)}
                  aria-labelledby={headingID}
                  aria-describedby={bodyID}
                  element={element}
                  ref={ref}
                  role="alertdialog"
                  style={styles}
                >
                  <AlertDialogHeader headingID={headingID} element={`${element}_HEADER`}>
                    {heading}
                  </AlertDialogHeader>
                  <AlertDialogBody bodyID={bodyID} element={`${element}_BODY`}>
                    {children}
                  </AlertDialogBody>
                  <AlertDialogFooter
                    destructive={destructive}
                    element={`${element}_FOOTER`}
                    onDismiss={onDismiss}
                    onDismissLabel={onDismissLabel}
                    onConfirm={onConfirm}
                    onConfirmLabel={onConfirmLabel}
                  />
                </Box>
              </ModalDialogOverlay>
            )
        )}
      </>
    );
  }
);

AlertDialog.displayName = 'AlertDialog';

AlertDialog.propTypes = {
  children: PropTypes.node.isRequired,
  destructive: PropTypes.bool,
  element: PropTypes.string,
  heading: PropTypes.string.isRequired,
  isOpen: PropTypes.bool.isRequired,
  onConfirm: PropTypes.func.isRequired,
  onConfirmLabel: PropTypes.string.isRequired,
  onDismiss: PropTypes.func.isRequired,
  onDismissLabel: PropTypes.string.isRequired,
};
