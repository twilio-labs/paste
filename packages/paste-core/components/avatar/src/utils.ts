import {
  IconSizeOptions,
  LineHeightOptions,
  FontSizeOptions,
  IconSize,
  LineHeight,
  FontSize,
} from '@twilio-paste/style-props';

export const getInitialsFromName = (fullname: string): string => {
  return fullname
    .split(' ')
    .map(name => name.slice(0, 1))
    .reduce((previous, currentInitial, index, initials) => {
      if (index === 0 || index === initials.length - 1) {
        return `${previous}${currentInitial}`;
      }
      return previous;
    }, '');
};

export const getCorrespondingLineHeightFromSizeToken = (size: IconSizeOptions): LineHeightOptions =>
  size.replace('sizeIcon', 'lineHeight') as LineHeightOptions;

export const getCorrespondingFontSizeFromSizeToken = (size: IconSizeOptions): FontSizeOptions => {
  switch (size) {
    case 'sizeIcon10':
    case 'sizeIcon20':
    case 'sizeIcon30':
    case 'sizeIcon40':
    case 'sizeIcon50':
    case 'sizeIcon60':
    default:
      return 'fontSize10';
    case 'sizeIcon70':
      return 'fontSize20';
    case 'sizeIcon80':
      return 'fontSize30';
    case 'sizeIcon90':
      return 'fontSize40';
    case 'sizeIcon100':
      return 'fontSize60';
    case 'sizeIcon110':
      return 'fontSize70';
  }
};

export const getComputedTokenNames = (size: IconSize): {lineHeight: LineHeight; fontSize: FontSize} => {
  if (Array.isArray(size)) {
    return {
      lineHeight: size.map(s => {
        if (s != null) {
          return getCorrespondingLineHeightFromSizeToken(s);
        }
        return null;
      }),
      fontSize: size.map(s => {
        if (s != null) {
          return getCorrespondingFontSizeFromSizeToken(s);
        }
        return null;
      }),
    };
  }
  return {
    lineHeight: getCorrespondingLineHeightFromSizeToken(size as IconSizeOptions),
    fontSize: getCorrespondingFontSizeFromSizeToken(size as IconSizeOptions),
  };
};
