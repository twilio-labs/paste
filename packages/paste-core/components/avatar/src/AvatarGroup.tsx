import { Stack } from "@twilio-paste/stack";
import * as React from "react";

import type { AvatarGroupProps } from "./types";
import { getGroupSpacing } from "./utils";

export const AvatarGroupContext = React.createContext<Omit<AvatarGroupProps, "children">>({} as AvatarGroupProps);

export const AvatarGroup = React.forwardRef<HTMLDivElement, AvatarGroupProps>(
  ({ variant, size, children, element = "AVATAR_GROUP" }, ref) => {
    return (
      <AvatarGroupContext.Provider value={{ variant, size }}>
        <Stack orientation="horizontal" spacing={getGroupSpacing(size, variant)} element={element} ref={ref}>
          {children}
        </Stack>
      </AvatarGroupContext.Provider>
    );
  },
);

AvatarGroup.displayName = "AvatarGroup";
