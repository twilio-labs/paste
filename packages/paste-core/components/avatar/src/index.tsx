import * as React from 'react';
import * as PropTypes from 'prop-types';
import {Text} from '@twilio-paste/text';
import {Box, safelySpreadBoxProps} from '@twilio-paste/box';
import {IconSize, isIconSizeTokenProp} from '@twilio-paste/style-props';
import {getComputedTokenNames, getInitialsFromName} from './utils';

export interface AvatarProps extends React.HTMLAttributes<'div'> {
  name: string;
  size: IconSize;
  src?: string;
}
const Avatar = React.forwardRef<HTMLDivElement, AvatarProps>(
  ({name, children, size = 'sizeIcon70', src, ...props}, ref) => {
    const computedTokenNames = getComputedTokenNames(size);
    if (src != null && name === undefined) {
      console.error('Paste Avatar: You must provide a name if you are displaying an image');
    }
    return (
      <Box
        {...safelySpreadBoxProps(props)}
        as="div"
        backgroundColor="colorBackgroundUser"
        borderRadius="borderRadiusCircle"
        overflow="hidden"
        ref={ref}
        size={size}
      >
        {src != null ? (
          <Box
            {...safelySpreadBoxProps(props)}
            as="img"
            alt={name}
            maxWidth="100%"
            src={src}
            size={size}
            title={name}
          />
        ) : (
          <Text
            as="abbr"
            display="block"
            fontSize={computedTokenNames.fontSize}
            fontWeight="fontWeightBold"
            lineHeight={computedTokenNames.lineHeight}
            textAlign="center"
            textDecoration="none"
            title={name}
          >
            {getInitialsFromName(name)}
          </Text>
        )}
      </Box>
    );
  }
);

Avatar.displayName = 'Avatar';
Avatar.propTypes = {
  size: isIconSizeTokenProp,
  src: PropTypes.string,
  name: PropTypes.string.isRequired,
};

export {Avatar};
