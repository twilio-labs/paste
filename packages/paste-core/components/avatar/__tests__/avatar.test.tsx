import * as React from 'react';
import {render, screen} from '@testing-library/react';
import axe from '../../../../../.jest/axe-helper';
import {Avatar} from '../src';
import {
  getCorrespondingLineHeightFromSizeToken,
  getCorrespondingFontSizeFromSizeToken,
  getComputedTokenNames,
  getInitialsFromName,
} from '../src/utils';

describe('Avatar', () => {
  describe('Utils', () => {
    describe('getInitialsFromName', () => {
      it('should return first and last initial from a name', () => {
        expect(getInitialsFromName('Philando Castile')).toEqual('PC');
        expect(getInitialsFromName('Tamir')).toEqual('T');
        expect(getInitialsFromName('Chimamanda Ngozi Adichie')).toEqual('CA');
        expect(getInitialsFromName('Super Long Name Example')).toEqual('SE');
      });
    });
    describe('getCorrespondingLineHeightFromSizeToken', () => {
      it('should return a line height to match icon size', () => {
        expect(getCorrespondingLineHeightFromSizeToken('sizeIcon10')).toEqual('lineHeight10');
        expect(getCorrespondingLineHeightFromSizeToken('sizeIcon20')).toEqual('lineHeight20');
        expect(getCorrespondingLineHeightFromSizeToken('sizeIcon30')).toEqual('lineHeight30');
        expect(getCorrespondingLineHeightFromSizeToken('sizeIcon40')).toEqual('lineHeight40');
        expect(getCorrespondingLineHeightFromSizeToken('sizeIcon50')).toEqual('lineHeight50');
        expect(getCorrespondingLineHeightFromSizeToken('sizeIcon60')).toEqual('lineHeight60');
        expect(getCorrespondingLineHeightFromSizeToken('sizeIcon70')).toEqual('lineHeight70');
        expect(getCorrespondingLineHeightFromSizeToken('sizeIcon80')).toEqual('lineHeight80');
        expect(getCorrespondingLineHeightFromSizeToken('sizeIcon90')).toEqual('lineHeight90');
        expect(getCorrespondingLineHeightFromSizeToken('sizeIcon100')).toEqual('lineHeight100');
        expect(getCorrespondingLineHeightFromSizeToken('sizeIcon110')).toEqual('lineHeight110');
      });
    });

    describe('getCorrespondingFontSizeFromSizeToken', () => {
      it('should return a font size to match icon size', () => {
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon10')).toEqual('fontSize10');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon20')).toEqual('fontSize10');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon30')).toEqual('fontSize10');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon40')).toEqual('fontSize10');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon50')).toEqual('fontSize10');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon60')).toEqual('fontSize10');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon70')).toEqual('fontSize20');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon80')).toEqual('fontSize30');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon90')).toEqual('fontSize40');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon100')).toEqual('fontSize60');
        expect(getCorrespondingFontSizeFromSizeToken('sizeIcon110')).toEqual('fontSize70');
      });
    });

    describe('getComputedTokenNames', () => {
      it('should handle single size values', () => {
        expect(getComputedTokenNames('sizeIcon50')).toEqual({fontSize: 'fontSize10', lineHeight: 'lineHeight50'});
      });
      it('should handle responsive size values', () => {
        expect(getComputedTokenNames(['sizeIcon50', 'sizeIcon100'])).toEqual({
          fontSize: ['fontSize10', 'fontSize60'],
          lineHeight: ['lineHeight50', 'lineHeight100'],
        });
      });
    });
  });

  describe('intials', () => {
    it('should render responsive css', (): void => {
      const {asFragment} = render(<Avatar size={['sizeIcon10', 'sizeIcon60', 'sizeIcon100']} name="Simon Taggart" />);
      expect(asFragment()).toMatchSnapshot();
    });
  });
  describe('image', () => {
    it('should render alt and src attributes', (): void => {
      render(<Avatar size="sizeIcon10" name="avatar example" src="/avatars/avatar2.png" />);
      expect(screen.getByRole('img').getAttribute('src')).toEqual('/avatars/avatar2.png');
      expect(screen.getByRole('img').getAttribute('alt')).toEqual('avatar example');
    });
    it('should render responsive css with an image', (): void => {
      const {asFragment} = render(
        <Avatar size={['sizeIcon30', 'sizeIcon40', 'sizeIcon90']} name="avatar example" src="/avatars/avatar2.png" />
      );
      expect(asFragment()).toMatchSnapshot();
    });
  });

  describe('accessibility', () => {
    it('should have no accessibility violations', async () => {
      const {container} = render(
        <>
          <Avatar size="sizeIcon10" name="Simon Taggart" />
          <Avatar size="sizeIcon10" name="avatar example" src="/avatars/avatar2.png" />
        </>
      );
      const results = await axe(container);
      expect(results).toHaveNoViolations();
    });
  });
});
