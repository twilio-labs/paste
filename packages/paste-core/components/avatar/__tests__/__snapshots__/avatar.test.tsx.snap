// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Avatar image should render responsive css with an image 1`] = `
<DocumentFragment>
  .emotion-1 {
  box-sizing: border-box;
  background-color: colorBackgroundUser;
  border-radius: borderRadiusCircle;
  overflow: hidden;
  width: sizeIcon30;
  height: sizeIcon30;
}

@media screen and (min-width:40em) {
  .emotion-1 {
    width: sizeIcon40;
    height: sizeIcon40;
  }
}

@media screen and (min-width:52em) {
  .emotion-1 {
    width: sizeIcon90;
    height: sizeIcon90;
  }
}

.emotion-0 {
  box-sizing: border-box;
  max-width: 100%;
  width: sizeIcon30;
  height: sizeIcon30;
}

@media screen and (min-width:40em) {
  .emotion-0 {
    width: sizeIcon40;
    height: sizeIcon40;
  }
}

@media screen and (min-width:52em) {
  .emotion-0 {
    width: sizeIcon90;
    height: sizeIcon90;
  }
}

<div
    class="emotion-1"
    data-paste-element="BOX"
    overflow="hidden"
  >
    <img
      alt="avatar example"
      class="emotion-0"
      data-paste-element="BOX"
      src="/avatars/avatar2.png"
      title="avatar example"
    />
  </div>
</DocumentFragment>
`;

exports[`Avatar intials should render responsive css 1`] = `
<DocumentFragment>
  .emotion-1 {
  box-sizing: border-box;
  background-color: colorBackgroundUser;
  border-radius: borderRadiusCircle;
  overflow: hidden;
  width: sizeIcon10;
  height: sizeIcon10;
}

@media screen and (min-width:40em) {
  .emotion-1 {
    width: sizeIcon60;
    height: sizeIcon60;
  }
}

@media screen and (min-width:52em) {
  .emotion-1 {
    width: sizeIcon100;
    height: sizeIcon100;
  }
}

.emotion-0 {
  margin: 0;
  padding: 0;
  color: colorText;
  font-size: fontSize10;
  line-height: lineHeight10;
  display: block;
  font-weight: fontWeightBold;
  text-align: center;
  -webkit-text-decoration: none;
  text-decoration: none;
}

@media screen and (min-width:40em) {
  .emotion-0 {
    font-size: fontSize10;
    line-height: lineHeight60;
  }
}

@media screen and (min-width:52em) {
  .emotion-0 {
    font-size: fontSize60;
    line-height: lineHeight100;
  }
}

<div
    class="emotion-1"
    data-paste-element="BOX"
    overflow="hidden"
  >
    <abbr
      class="emotion-0"
      color="colorText"
      data-paste-element="TEXT"
      display="block"
      font-size="fontSize10,fontSize10,fontSize60"
      font-weight="fontWeightBold"
      text-decoration="none"
      title="Simon Taggart"
    >
      ST
    </abbr>
  </div>
</DocumentFragment>
`;
